### DISCUSSION TITLE
Can anyone teach me how to use it on openwrt, please

### DISCUSSION THREAD

**Discussion by tao593** [AdguardTeam/AdGuardHome#3574] [Q&A]
root@OpenWrt:/usr/bin#./AdGuardHome -s install --no-check-update
panic: qtls.ClientHelloInfo doesn't match

goroutine 1 [running]:
github.com/marten-seemann/qtls-go1-15.init.0()
	github.com/marten-seemann/qtls-go1-15@v0.1.1/unsafe.go:20 +0x1b4

**Comment by tao593** [comment]
OpenWrt 21.02.0 r16279-5cc0535800 / LuCI openwrt-21.02 branch git-21.231.26241-422c175

**Comment by tao593** [comment]
#!/bin/sh /etc/rc.common

PROG=/usr/bin/AdGuardHome
USE_PROCD=1

# starts after network starts
START=21
# stops before networking stops
STOP=89

start_service() {
  procd_open_instance
  procd_set_param command "$PROG" -c /etc/adguardhome.yaml -w /tmp/adguardhome --no-check-update
  procd_set_param stdout 1
  procd_set_param stderr 1
  procd_close_instance
}

**Comment by tao593** [comment]
#
# This is free software, licensed under the GNU General Public License v2.
# See /LICENSE for more information.
#

include $(TOPDIR)/rules.mk

PKG_NAME:=adguardhome
PKG_VERSION:=0.104.3
PKG_RELEASE:=1

PKG_SOURCE_PROTO:=git
PKG_SOURCE_VERSION:=v$(PKG_VERSION)
PKG_SOURCE_URL:=https://github.com/AdguardTeam/AdGuardHome
PKG_MIRROR_HASH:=9051c08ebefccd918cad9b487d2d3b2c4b276ac71f16706c2ae8ee2a37ba9d03

PKG_LICENSE:=GPL-3.0-only
PKG_LICENSE_FILES:=LICENSE.txt
PKG_MAINTAINER:=Dobroslaw Kijowski <dobo90@gmail.com>

PKG_BUILD_DEPENDS:=golang/host node/host packr/host
PKG_BUILD_PARALLEL:=1
PKG_USE_MIPS16:=0

GO_PKG:=github.com/AdguardTeam/AdGuardHome
GO_PKG_EXCLUDES:=dhcpd/standalone
GO_PKG_LDFLAGS_X:=main.version=$(PKG_VERSION) main.channel=release main.goarm=$(GO_ARM)

include $(INCLUDE_DIR)/package.mk
include ../../lang/golang/golang-package.mk

define Package/adguardhome
	SECTION:=net
	CATEGORY:=Network
	TITLE:=Network-wide ads and trackers blocking DNS server
	URL:=https://github.com/AdguardTeam/AdGuardHome
	DEPENDS:=$(GO_ARCH_DEPENDS) +ca-bundle
endef

define Package/adguardhome/conffiles
/etc/adguardhome.yaml
endef

define Package/adguardhome/description
Free and open source, powerful network-wide ads and trackers blocking DNS server.
endef

define Build/Compile
	( \
		pushd $(PKG_BUILD_DIR) ; \
		npm --prefix client ci ; \
		npm --prefix client run build-prod ; \
		packr -z -v -i internal ; \
		popd ; \
		$(call GoPackage/Build/Compile) ; \
	)
endef

define Package/adguardhome/install
	$(call GoPackage/Package/Install/Bin,$(1))
	$(INSTALL_DIR) $(1)/etc/init.d
	$(INSTALL_BIN) ./files/adguardhome.init $(1)/etc/init.d/adguardhome
endef

$(eval $(call GoBinPackage,adguardhome))
$(eval $(call BuildPackage,adguardhome))

**Comment by tao593** [comment]
What went wrong

**Comment by ainar-g** [comment]
Hello.  This looks like a custom build of AGH, and all sorts of things could go wrong.  Please use the official `Makefile` and build scripts to build your AGH or just use a released version.

**Reply by tao593** [reply]
https://github.com/openwrt/packages/issues/16594
In this discussion, it is enough to replace the binary file. I tried it and it worked.

**Comment by ali27868** [comment]
https://forum.openwrt.org/t/openwrt-adguard-home-101-dnsmasq/110864

Please pay attention to the script in the same thread at https://forum.openwrt.org/t/openwrt-adguard-home-101-dnsmasq/110864/39

For me worked the manual install as per the first link (the only difference is I did it on external USB storage) + dnsmasq / DHCP settings from the second link.

External storage set-up is well described here: https://www.youtube.com/watch?v=5UjCT99udm0 (just skipping the AdGuard part from this video)
